src/model/project.ts

```diff
@@ -0,0 +1,102 @@
+/**
+ * @license
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+import {Warning} from '../warning/warning';
+
+import {Document, FeatureKinds} from './document';
+import {Feature} from './feature';
+
+export class Project {
+  private _rootDocuments: Set<Document>;
+  private _toplevelWarnings: Warning[];
+
+  constructor(rootDocuments: Iterable<Document>, warnings: Warning[]) {
+    this._rootDocuments = new Set(rootDocuments);
+    this._toplevelWarnings = warnings;
+
+    // This is a performance optimization. We only need a set of documents such
```

Comment:
The optimization is in the expense of duplicate work when querying, not in analyzing.